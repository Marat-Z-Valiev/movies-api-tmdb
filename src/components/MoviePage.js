import React, {useState, useEffect} from "react";
import axios from "axios";
import PropTypes from "prop-types";
import MovieStyled from "./styles/MovieStyled";
import PeopleContainer from "./PeopleContainer";
import Video from "./Video";
import Spinner from "./Spinner";

const MoviePage = ({match}) => {
	const movieId = match.params.id;
	let [movie, setMovie] = useState({});
	let [people, setPeople] = useState([]);
	let [isLoading, setIsLoading] = useState(true);

	const getMovie = async () => {
		await axios
			.get(
				`https://api.themoviedb.org/3/movie/${movieId}?api_key=${process.env.API_KEY}&language=en-US`
			)
			.then((response) => {
				setMovie((movie = response.data));
				setIsLoading((isLoading = false));
			})
			.catch((err) => console.log(`this is error ${err}`));
	};

	const getActors = async () => {
		await axios
			.get(
				`https://api.themoviedb.org/3/movie/${movieId}/credits?api_key=${process.env.API_KEY}`
			)
			.then((response) => {
				setPeople((people = response.data.cast));
				setIsLoading((isLoading = false));
			})
			.catch((err) => console.log(`this is error ${err}`));
	};

	useEffect(() => {
		getMovie();
	}, []);

	useEffect(() => {
		getActors();
	}, []);

	const {
		title,
		poster_path,
		vote_average,
		release_date,
		runtime,
		overview,
	} = movie;

	return (
		<>
			{isLoading ? (
				<Spinner />
			) : (
				<>
					<h1 style={{textAlign: "center"}}>{title}</h1>
					<MovieStyled>
						<img
							src={`https://image.tmdb.org/t/p/w342/${poster_path}`}
							alt={title}
						/>
						<div className="movie-details">
							<div className="movie-stats">
								<h2>
									<i className="far fa-calendar-alt fa-lg" />
									{release_date}
								</h2>
								<h2>
									<i className="fas fa-star fa-lg" />
									{vote_average == 0 ? "" : `Popularity ${vote_average} / 10`}
								</h2>
								<h2>
									<i className="far fa-clock fa-lg" />
									{runtime} mins
								</h2>
							</div>
							<div className="overview">
								<p>{overview}</p>
							</div>
						</div>
					</MovieStyled>
					{people.length ? (
						<>
							<h1 style={{textAlign: "center"}}>Cast</h1>
							<PeopleContainer
								people={people.slice(0, 6)}
								isLoading={isLoading}
							/>
						</>
					) : (
						""
					)}
					<Video movieId={movieId}></Video>
				</>
			)}
		</>
	);
};

MoviePage.propTypes = {
	match: PropTypes.object.isRequired,
};

export default MoviePage;
