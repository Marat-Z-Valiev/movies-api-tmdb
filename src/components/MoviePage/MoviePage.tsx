import React, {useEffect} from "react";
import {useDispatch, useSelector} from "react-redux";
import PropTypes from "prop-types";
import MovieStyled from "./MovieStyled";
import PeopleContainer from "../PeopleContainer/PeopleContainer";
import Video from "../Video/Video";
import Spinner from "../Spinner/Spinner";
import Error from "../Error/Error";
import {fetchMovie, movieSelector} from "../../store/slices/movie";
import {fetchMovieCast, castSelector} from "../../store/slices/movieCast";

const MoviePage = ({match}: any) => {
	const movieId = match.params.id;
	const dispatch = useDispatch();
	const {loading, movie, hasErrors} = useSelector(movieSelector);
	const {cast} = useSelector(castSelector);
	useEffect(() => {
		dispatch(fetchMovie(movieId));
	}, [dispatch]);

	useEffect(() => {
		dispatch(fetchMovieCast(movieId));
	}, [dispatch]);

	const {
		title,
		poster_path,
		vote_average,
		release_date,
		runtime,
		overview,
	} = movie;

	return (
		<>
			{hasErrors ? <Error /> : ""}
			{loading ? (
				<Spinner />
			) : (
				<>
					<h1 style={{textAlign: "center"}}>{title}</h1>
					<MovieStyled>
						<img
							src={`https://image.tmdb.org/t/p/w342/${poster_path}`}
							alt={title}
						/>
						<div className="movie-details">
							<div className="movie-stats">
								<h2>
									<i className="far fa-calendar-alt fa-lg" />
									{release_date}
								</h2>
								<h2>
									<i className="fas fa-star fa-lg" />
									{vote_average == 0 ? "" : `Popularity ${vote_average} / 10`}
								</h2>
								<h2>
									<i className="far fa-clock fa-lg" />
									{runtime} mins
								</h2>
							</div>
							<div className="overview">
								<p>{overview}</p>
							</div>
						</div>
					</MovieStyled>
					{cast.length ? (
						<>
							<h1 style={{textAlign: "center"}}>Cast</h1>
							<PeopleContainer people={cast.slice(0, 6)} isLoading={loading} />
						</>
					) : (
						""
					)}
					<Video movieId={movieId}></Video>
				</>
			)}
		</>
	);
};

MoviePage.propTypes = {
	match: PropTypes.object.isRequired,
};

export default MoviePage;
