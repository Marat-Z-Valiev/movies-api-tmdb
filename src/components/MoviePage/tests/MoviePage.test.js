import React from "react";
import {shallow, configure} from "enzyme";
import toJSON from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import MoviePage from "../MoviePage";

configure({adapter: new Adapter()});
jest.mock("axios");
describe("MoviePage", () => {
	it("should render corretly", () => {
		const props = {
			match: {
				params: {
					id: 1,
				},
			},
		};
		const wrapper = shallow(<MoviePage {...props} />);
		expect(toJSON(wrapper)).toMatchSnapshot();
	});
});
