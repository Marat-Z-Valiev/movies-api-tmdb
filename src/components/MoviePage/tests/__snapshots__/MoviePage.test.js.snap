// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePage should render corretly 1`] = `
<Fragment>
  <Spinner />
</Fragment>
`;
